File: jhyun/CodeTree/LR Technique/씨_오_더블유.java
import java.util.*;
import java.io.*;

public class Main {
    static final int MAX_N = 100000;

    static long L[] = new long[MAX_N];
    static long R[] = new long[MAX_N];
    static long ans;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        String word = br.readLine();

        //L[i] = 0번부터 i번 문자 C의 개수
        L[0] = (word.charAt(0) == 'C' ? 1 : 0);
        for (int i = 1; i < n; i++) {
            L[i] = L[i - 1] + (word.charAt(i) == 'C' ? 1 : 0);
        }

        //L[i] = i번부터 n-1번 문자 중 W의 개수
        R[n - 1] = (word.charAt(n - 1) == 'W' ? 1 : 0);
        for (int i = n - 2; i >= 0; i--) {
            R[i] = R[i + 1] + (word.charAt(i) == 'W' ? 1 : 0);
        }

        //각 위치 i에 대해 word[i]가 O라면
        //i보다 왼쪽에 있는 C의 개수인 L[i-1] &
        //i보다 오른쪽에 있는 W의 개수인 R[i+1]의 곱이
        //i를 중심으로 하여 COW 조합을 만들어 내는 숫자
        for (int i = 1; i < n - 1; i++) {
            if (word.charAt(i) == 'O') {
                ans += L[i - 1] * R[i + 1];
            }
        }
        System.out.print(ans);
    }
}

File: Park/CodeTree/LR Technique/씨 오 더블유.py
# 변수 선언 및 입력 받기
n = int(input())  # 문자열 길이
word = input()    # 문자열 입력

# L과 R 배열을 0으로 초기화
L = [0] * n  # L[i]는 0부터 i까지 'C'의 개수
R = [0] * n  # R[i]는 i부터 n-1까지 'W'의 개수
ans = 0      # 'COW' 패턴의 개수를 저장할 변수

# L 배열 계산: 각 인덱스까지 'C'의 누적 개수
if word[0] == 'C':
    L[0] = 1
for i in range(1, n):
    L[i] = L[i - 1] + (1 if word[i] == 'C' else 0)

# R 배열 계산: 각 인덱스부터 끝까지 'W'의 누적 개수
if word[n - 1] == 'W':
    R[n - 1] = 1
for i in range(n - 2, -1, -1):
    R[i] = R[i + 1] + (1 if word[i] == 'W' else 0)

# 'O'가 있는 위치에서 왼쪽의 'C'와 오른쪽의 'W'의 곱을 계산
for i in range(1, n - 1):
    if word[i] == 'O':
        ans += L[i - 1] * R[i + 1]

print(ans)
